src/mgmt/srv_target.c

```diff
@@ -657,3 +658,194 @@ ds_mgmt_hdlr_tgt_destroy(crt_rpc_t *td_req)
 	td_out->td_rc = rc;
 	crt_reply_send(td_req);
 }
+
+/*
+ * Return a URI string, like "ofi+sockets://192.168.1.70:44821", that callers
+ * are responsible for freeing with D_FREE.
+ */
+static int
+create_tag_uri(const char *base_uri, int i, char **uri)
+{
+	char   *type = "ofi+sockets:";
+	char   *u;
+	char   *p;
+	int	port;
+
+	/* TODO: Support other URI types. */
+	if (strncmp(base_uri, type, strlen(type)) != 0) {
+		D_ERROR("URI %s type not supported\n", base_uri);
+		return -DER_NOSYS;
+	}
+
+	/* Locate the ":" between the host and the port. */
+	p = strrchr(base_uri, ':');
+	if (p == NULL) {
+		D_ERROR("no ':' in base URI %s\n", base_uri);
+		return -DER_INVAL;
+	}
+
+	/* Calculate the port. */
+	port = atoi(p + 1) + i;
+	if (port > 65535 /* maximal port number */) {
+		D_ERROR("port %d out of range\n", port);
+		return -DER_INVAL;
+	}
+
+	/* Print the base URI with the port into u. */
+	D_ASPRINTF(u, "%.*s%d", (int)(p + 1 - base_uri), base_uri, port);
+	if (u == NULL)
+		return -DER_NOMEM;
+
+	*uri = u;
+	return 0;
+}
+
+static int
+add_server(crt_group_t *group, struct server_entry *server)
+{
+	int i;
+
+	D_DEBUG(DB_MGMT, "rank=%u uri=%s nctxs=%u\n", server->se_rank,
+		server->se_uri, server->se_nctxs);
+
+	for (i = 0; i < server->se_nctxs; i++) {
+		crt_node_info_t	info;
+		int		rc;
+
+		rc = create_tag_uri(server->se_uri, i, &info.uri);
+		if (rc != 0)
+			return rc;
+		rc = crt_group_node_add(group, server->se_rank, i, info);
+		D_FREE(info.uri);
+		if (rc != 0) {
+			D_ERROR("failed to add rank=%u tag=%d uri=%s: %d\n",
+				server->se_rank, i, info.uri, rc);
+			return rc;
+		}
+	}
+
+	return 0;
+}
+
+static int
+remove_server(crt_group_t *group, struct server_entry *server)
+{
+	int rc;
+
+	D_DEBUG(DB_MGMT, "rank=%u uri=%s nctxs=%u\n", server->se_rank,
+		server->se_uri, server->se_nctxs);
+
+	rc = crt_group_rank_remove(group, server->se_rank);
+	if (rc != 0)
+		D_ERROR("failed to remove rank=%u: %d\n", server->se_rank, rc);
+
+	return rc;
+}
+
+/* State of the local system map (i.e., the CaRT PG membership) */
+static uint32_t sys_map_version;
+
+int
+ds_mgmt_tgt_map_update_pre_forward(crt_rpc_t *rpc, void *arg)
+{
+	struct mgmt_tgt_map_update_in  *in = crt_req_get(rpc);
+	struct server_entry	       *servers = in->tm_servers.ca_arrays;
+	crt_group_t		       *group = crt_group_lookup(NULL);
+	d_rank_t			self_rank;
+	d_rank_list_t		       *ranks;
+	int				i;
+	int				rc;
+
+	if (in->tm_map_version <= sys_map_version)
+		return 0;
+
+	rc = crt_group_rank(group, &self_rank);
+	if (rc != 0) {
+		D_DEBUG(DB_MGMT, "self rank unknown: %d\n", rc);
+		return rc;
+	}
+
+	rc = crt_group_ranks_get(group, &ranks);
+	if (rc != 0) {
+		D_ERROR("failed to get existing ranks: %d\n", rc);
+		return rc;
+	}
+#if 1
```

Comment:
sorry, what is the intention for # if 1?

Reply:
These are for debugging purposes only; let me remove them.